### PR TITLE
rate limit gregor push state calls

### PR DESCRIPTION
**Pull request by mmaxim**

### CODE REVIEW

File: go/service/gregor.go [hunk 1]
```diff
@@ -1298,6 +1318,7 @@ func (g *gregorHandler) Reset() error {
 	g.Shutdown()
 	g.setFirstConnect(true)
 	g.shutdownGregorClient(context.TODO())
+	g.pushStateCh <- nil // clear any pending push state calls
```
Comment (joshblum) [inline go/service/gregor.go]: close on shutdown?

File: go/service/gregor.go [hunk 2]
```diff
@@ -521,21 +525,37 @@ func (g *gregorHandler) iterateOverFirehoseHandlers(f func(h libkb.GregorFirehos
 	g.firehoseHandlers = freshHandlers
 }
 
+func (g *gregorHandler) pushStateThread() {
+	var lastPush func()
+	var lastTime time.Time
+	for push := range g.pushStateCh {
+		lastPush = push
+		if time.Since(lastTime) > time.Second {
```
Comment (joshblum) [inline go/service/gregor.go]: if this is false we just lose the update, seems wrong

Comment (mmaxim) [reply go/service/gregor.go]: I believe this is handled in my latest version

File: go/service/gregor.go (hunk 1 above)
Comment (mmaxim) [reply go/service/gregor.go]: I'm not sure how to get it going again given current construction.

Comment (mmaxim) [reply go/service/gregor.go]: Nvm, made a change

File: go/service/gregor.go [hunk 3]
```diff
@@ -524,20 +524,20 @@ func (g *gregorHandler) iterateOverFirehoseHandlers(f func(h libkb.GregorFirehos
 }
 
 func (g *gregorHandler) pushStateThread(shutdownCh chan struct{}) {
```
Comment (joshblum) [inline go/service/gregor.go]: maybe rename to `debounceNewData`

File: go/service/gregor.go [hunk 4]
```diff
@@ -549,22 +549,29 @@ func (g *gregorHandler) pushStateThread(shutdownCh chan struct{}) {
...
+	// Only send this state update on reception of new data, not a reconnect since we will
+	// be sending that on a different code path altogether (see OnConnect).
+	if g.badger != nil && r != keybase1.PushReason_RECONNECTED {
+		g.badger.PushState(context.Background(), s)
+	}
+}
+
+func (g *gregorHandler) pushState(r keybase1.PushReason) {
+	switch r {
+	case keybase1.PushReason_RECONNECTED, keybase1.PushReason_NONE:
```
Comment (joshblum) [inline go/service/gregor.go]: when does `PushReason_NONE` happen? does it ever? why not debounce?

File: go/service/gregor.go [hunk 5]
```diff
@@ -521,7 +523,33 @@ func (g *gregorHandler) iterateOverFirehoseHandlers(f func(h libkb.GregorFirehos
...
+	for {
+		select {
+		case <-g.pushStateCh:
+			shouldSend = true
+			if time.Since(lastTime) > dur {
+				trigger()
+			}
+		case <-time.After(dur):
+			trigger()
+		case <-shutdownCh:
```
Comment (joshblum) [inline go/service/gregor.go]: this never gets reinitialized on reconnect

File: go/service/gregor.go (hunk 4 above)
Comment (mmaxim) [reply go/service/gregor.go]: Unused, and I have `NEW_DATA` above as the reason, didn't want to muddle it.

File: go/service/gregor.go (hunk 5 above)
Comment (mmaxim) [reply go/service/gregor.go]: It does in `Connect`

Comment (joshblum) [reply go/service/gregor.go]: oops missed this, lgtm!

**Pull request merged**